nucypher/policy/conditions/lingo.py

```diff
@@ -3,68 +3,117 @@
+        return value.to_dict()
 
-    @classmethod
-    def from_dict(cls, data: OperatorDict) -> "Operator":
-        cls.validate(data)
-        instance = cls(_operator=data["operator"])
-        return instance
+    def _deserialize(self, value, attr, data, **kwargs):
+        condition = deserialize_condition_lingo(value)
+        return condition
 
-    @classmethod
-    def from_json(cls, data) -> 'Operator':
-        data = json.loads(data)
-        instance = cls.from_dict(data)
-        return instance
 
-    def to_json(self) -> str:
-        data = self.to_dict()
-        json_data = json.dumps(data)
-        return json_data
+#
+# CONDITION = BASE_CONDITION | COMPOUND_CONDITION
+#
+# BASE_CONDITION = {
+#     // ..
+# }
+#
+# COMPOUND_CONDITION = {
+#     "operator": OPERATOR,
+#     "operands": [CONDITION*]
+# }
 
-    @classmethod
-    def validate(cls, data: Dict) -> None:
-        try:
-            _operator = data["operator"]
-        except KeyError:
-            raise InvalidLogicalOperator(f"Invalid operator data: {data}")
 
-        if _operator not in cls.OPERATORS:
-            raise InvalidLogicalOperator(f"{_operator} is not a valid operator")
+class CompoundAccessControlCondition(AccessControlCondition):
+    AND_OPERATOR = "and"
+    OR_OPERATOR = "or"
+    OPERATORS = (AND_OPERATOR, OR_OPERATOR)
+
+    class Schema(CamelCaseSchema):
+        SKIP_VALUES = (None,)
+        name = fields.Str(required=False)
+        operator = fields.Str(required=True, validate=validate.OneOf(["and", "or"]))
+        operands = fields.List(
+            _ConditionsField, required=True, validate=validate.Length(min=2)
+        )
+
+        @post_load
+        def make(self, data, **kwargs):
+            return CompoundAccessControlCondition(**data)
+
+    def __init__(
+        self, operator: str, operands: List[Condition], name: Optional[str] = None
+    ):
+        """
+        COMPOUND_CONDITION = {
+            "operator": OPERATOR,
+            "operands": [CONDITION*]
+        }
+        """
+        if operator not in self.OPERATORS:
+            raise InvalidLogicalOperator(f"{operator} is not a valid operator")
+        self.operator = operator
+        self.operands = operands
+        self.name = name
+        self.id = md5(bytes(self)).hexdigest()[:6]
+
+    def __repr__(self):
+        return f"Operator={self.operator} (NumOperands={len(self.operands)}), id={self.id})"
+
+    def verify(self, *args, **kwargs) -> Tuple[bool, Any]:
+        values = []
+        overall_result = True if self.operator == self.AND_OPERATOR else False
+        for condition in self.operands:
+            current_result, current_value = condition.verify(*args, **kwargs)
+            values.append(current_value)
+            if self.operator == self.AND_OPERATOR:
+                overall_result = overall_result and current_result
+                # short-circuit check
+                if overall_result is False:
```

Comment:
Because of the short-circuiting, if any "and" condition returns `False` the subsequent conditions no longer need to be evaluated, because it doesn't matter whether the other conditions return True or False.

Consequently, if `condition_1` is False, we don't evaluate `condition_2`, `condition_3`, or `condition_4` (remember these are mocks whose results are set as part of the test, but in actuality `verify(...)` executes the condition code and returns True/False and the value obtained for comparison). Therefore, we don't have a value to return for those that aren't executed.

The conditions are evaluated in order, so `[1]` is the only actual result value we have., and corresponds to the first condition.

I debated returning `[1, None, None, None]` to make it explicit that the other conditions were not executed, but that seemed wasteful/superfluous to me. We can revisit this if the prevailing sentiment disagrees with me. 

The other thing is that the values returned, aren't actually used other than in tests. `ConditionLingo.eval(...)` which is the call made when evaluating conditions for CBD only returns the overall result True/False, but not the values returned by the conditions. In the future, we may decide that the values could be useful feedback to users, but at the moment we don't do anything with them - see https://github.com/derekpierre/nucypher/blob/prefix-conditions/nucypher/policy/conditions/lingo.py#L266

```python
ConditionsLingo:
    def eval(self, *args, **kwargs) -> bool:
        result, _ = self.condition.verify(*args, **kwargs)
        return result
 ```

Reply:
Got it! Thanks